### PR TITLE
[REEF-863] Test ActiveContext.SendMessage()

### PR DESCRIPTION
**Pull request by afchung**
This addressed the issue by
  * Adding a context message portion to the driver-evaluator messaging test.
  * Fixing typos dealing with ContextMessages.
  * Fixing ContextMessageClr2Java where it calls the wrong interop method in Java.

JIRA:
  [REEF-863](https://issues.apache.org/jira/browse/REEF-863)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+                return Optional<ContextMessage>.Of(
+                    ContextMessage.From("Context", ByteUtilities.StringToByteArrays(MessageSend)));
+            }
+        }
+
+        public void OnNext(byte[] messageBytes)
+        {
+            var message = ByteUtilities.ByteArraysToString(messageBytes);
+            Logger.Log(Level.Info, MessageReceivedFromDriverLog);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs]: Shouldn't you only log this when the message is actually the right message?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs [hunk 2]
```diff
@@ -32,60 +32,96 @@
...
-            eval.SubmitContextAndTask(contextConfiguration, taskConfiguration);
+            eval.SubmitContextAndService(contextConfiguration, serviceConfiguration);
         }
 
         public void OnNext(ITaskMessage taskMessage)
         {
-            string msgReceived = ByteUtilities.ByteArraysToString(taskMessage.Message);
+            var msgReceived = ByteUtilities.ByteArraysToString(taskMessage.Message);
+
+            Logger.Log(Level.Info, DriverReceivedTaskMessageLog);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: Shouldn't this only be logged if it is the right message?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs [hunk 3]
```diff
@@ -28,27 +27,38 @@
...
-        public HelloService Service { get; set; }
-
         public Optional<TaskMessage> Message
         {
             get
-            {              
-                TaskMessage defaultTaskMessage = TaskMessage.From(
+            {
+                var defaultTaskMessage = TaskMessage.From(
                     "messagingSourceId",
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: Shouldn't this be a constant we can verify on the other side?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs [hunk 4]
```diff
@@ -59,44 +69,33 @@ public Optional<TaskMessage> Message
...
         public byte[] Call(byte[] memento)
         {
-            Console.WriteLine("Hello, CLR TaskMsg!");
-            Thread.Sleep(5 * 1000);
-            return null;
-        }
+            while (!(_messageManager.IsTaskMessageSent && _messageManager.IsContextMessageSent))
+            {
+                // Do not exit until task and context messages are both sent.
+                Thread.Sleep(5000);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: This waits at most 5 seconds after the last message is sent, right? Can this be done with an interrupt instead? Such that we don't delay test completion by so much?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs [hunk 5]
```diff
@@ -59,44 +69,33 @@ public Optional<TaskMessage> Message
...
 
-            [Inject]
-            public MessagingDriverMessageHandler(MessageTask task)
+            var message = ByteUtilities.ByteArraysToString(value.Message.Value);
+            if (!message.Equals(MessageDriver.Message))
             {
-                _parentTask = task;
+                Exceptions.Throw(new Exception("Unexpected driver message: " + message), "Unexpected driver message received: " + message, Logger);
             }
+        }
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: This method does not log success. Is this on purpose? It seems the other way around in all the other methods.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs]: Shouldn't this only log if the right message was received?

**Comment by markusweimer** [comment]
I've done a pass, focused on the test itself, not the code that makes it pass.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs (hunk 5 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: The `MessageReceivedFromDriverLog` is the success message. If the message is incorrect, we get a FailedEvaluator on the Exception, which fails the test. I will move the order such that it only logs on the correct message.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs (hunk 3 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: I don't think this is actually implemented in `ITaskMessage`.

Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: Opened REEF-1385 and added TODO.

**Comment by afchung** [comment]
@markusweimer Addressed your comments, please have another look. Thanks!

File: lang/cs/Org.Apache.REEF.Common/Context/ContextMessage.cs [hunk 6]
```diff
@@ -43,7 +43,7 @@ public byte[] Bytes
         }
 
         /// <summary>
-        /// construt a new new ContextMessage with the given content.
+        /// construct a new new ContextMessage with the given content.
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextMessage.cs]: two "new"

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Tests.Functional.Messaging
+{
+    /// <summary>
+    /// A Context event handler class that generates context messages for the Driver 
+    /// and receives context messages from the Driver for testing.
+    /// </summary>
+    public sealed class MessageContext : IContextMessageHandler, IContextMessageSource
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs]: It is not a Context, so may be rename it as ContextMessgeSourceAndHandler or just ContextMessgeHandler.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs [hunk 8]
```diff
@@ -32,60 +32,111 @@
...
+    public sealed class MessageDriver :
         IObserver<IAllocatedEvaluator>, 
         IObserver<ITaskMessage>, 
+        IObserver<IContextMessage>,
         IObserver<IRunningTask>, 
-        IObserver<IDriverStarted>
+        IObserver<IDriverStarted>,
+        IObserver<IActiveContext>
     {
         public const int NumberOfEvaluator = 1;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: private

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs [hunk 9]
```diff
@@ -32,60 +32,111 @@
...
 
         public const string Message = "MESSAGE::DRIVER";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs [hunk 10]
```diff
@@ -32,60 +32,111 @@
...
 
-        private static readonly Logger LOGGER = Logger.GetLogger(typeof(MessageDriver));
+        public const string SendingMessageToTaskLog = "Sending message to Task.";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs [hunk 11]
```diff
@@ -32,60 +32,111 @@
...
+        public const string DriverReceivedTaskMessageLog = "Driver received message from Task.";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs [hunk 12]
```diff
@@ -32,60 +32,111 @@
...
+
+        public const string SendingMessageToContextLog = "Sending message to Context.";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs [hunk 13]
```diff
@@ -32,60 +32,111 @@
...
+        public const string DriverReceivedContextMessageLog = "Driver received message from Context.";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs [hunk 14]
```diff
@@ -16,39 +16,52 @@
...
+    /// </summary>
+    public class MessageTask : ITask, ITaskMessageSource, IDriverMessageHandler
     {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MessageTask));
+
         public const string MessageSend = "MESSAGE:TASK";
 
-        private static readonly Logger LOGGER = Logger.GetLogger(typeof(MessageTask));
+        // TODO[JIRA REEF-1385]: Check the MessageTaskSourceID on the Driver side.
+        public const string MessageTaskSourceId = "MessageTaskSourceId";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs [hunk 15]
```diff
@@ -16,39 +16,52 @@
...
+
+        public const string MessageSentToDriverLog = "Message sent to Driver from Task.";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs [hunk 16]
```diff
@@ -16,39 +16,52 @@
...
+        public const string MessageReceivedFromDriverLog = "Message received from Driver in Task.";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: internal

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/TestEvaluatorMessaging.cs [hunk 17]
```diff
@@ -15,15 +15,10 @@
 // specific language governing permissions and limitations
 // under the License.
 
-using System;
 using System.Collections.Generic;
-using System.Runtime.InteropServices;
+using System.Security.Policy;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/TestEvaluatorMessaging.cs]: Not used

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs [hunk 18]
```diff
@@ -16,39 +16,52 @@
...
-                    "messagingSourceId",
-                    ByteUtilities.StringToByteArrays(MessageSend + " generated at " + DateTime.Now.ToString(CultureInfo.InvariantCulture)));
-                LOGGER.Log(Level.Info, "Message is sent back from task to driver:" + defaultTaskMessage.Message);
+            {
+                var defaultTaskMessage = TaskMessage.From(
+                    MessageTaskSourceId,
+                    ByteUtilities.StringToByteArrays(MessageSend));
+                    
+                Logger.Log(Level.Info, MessageSentToDriverLog);
+                _messageManager.OnTaskMessageSent();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: You expect TaskMessage sent first, then release task, right? In this sequence, would it possible if task has completed but task message has not been received?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageManager.cs [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+using System.Threading;
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Tests.Functional.Messaging
+{
+    /// <summary>
+    /// A test helper class that keeps track of whether context/task messages 
+    /// have been sent.
+    /// </summary>
+    public sealed class MessageManager
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageManager.cs]: Would you like to rename it as Message MessageEventController  or something with Test in the name?

**Comment by jwang98052** [comment]
I completed a pass.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs (hunk 8 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: I don't think this matters, provided that this is a test class.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs (hunk 9 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: I don't think this matters, provided that this is a test class.

Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageDriver.cs]: Same for others.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs (hunk 18 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageTask.cs]: No, I wait for all 3 events -- context message sent, task message sent, and message received from driver below.

File: lang/cs/Org.Apache.REEF.Common/Context/ContextMessage.cs (hunk 6 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextMessage.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageManager.cs (hunk 19 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageManager.cs]: Done

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs (hunk 7 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/MessageContext.cs]: Done.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/TestEvaluatorMessaging.cs (hunk 17 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Messaging/TestEvaluatorMessaging.cs]: Done.

**Comment by afchung** [comment]
@jwang98052 I've addressed your comments, please have another look. Thanks!

**Comment by jwang98052** [comment]
LGTM

**Comment by afchung** [comment]
@jwang98052 or @markusweimer, is this ready to be merged? Thanks!

**Comment by jwang98052** [comment]
+1 for merge.

**Comment by markusweimer** [comment]
LGTM, +1 for merge

**Comment by jwang98052** [comment]
I will test and merge.

**Pull request closed without merging**